### DISCUSSION TITLE
About subscribing to topics using wildcards

### DISCUSSION THREAD

**Discussion by YYYYWD** [emqx/emqx#8276] [Q&A]
If I subscribe to a topic using wildcards, do I receive a large number of reserved messages under topics that match wildcards? Should this be limited?
For example, I use _testtopic/#_ to subscribe the topics, so broker need to scan the storage for all retained messages under the subject that matches this wildcard, right? Doesn't this have a big impact on performance?

**Comment by zmstone** [comment]
Hi @YYYYWD 
Yes, it does impact system performance and the client process may get killed due to large heap (RAM) size usage.
It's probably not a bit impact if you do not have many such clients though.

We have introduced rate limiting to retainer in EMQX v5.0.

**Reply by YYYYWD** [reply]
That means that under the mqtt3.11 protocol, this problem cannot be solved very well. 
If you have multiple brockers in a distributed environment, how does your code work? Every time you find that client uses a wildcard subscription, go to KV store to scan all eligible topics? Or do you maintain a global topic trie tree for every broker?

**Reply by zmstone** [reply]
> 

By `v5.0` I meant EMQX version 5, but not MQTT protocol spec version.
The retained messages are replicated to ALL nodes in the cluster.
Retained messages are not stored in trie, however, the trie is also replicated to ALL nodes in the cluster.

**Reply by YYYYWD** [reply]
Do you mean that there is the same trie tree in different nodes, and when you encounter a wildcard subscription, you search for the trie tree, get all the matching topics, and then search the kv store to search for the corresponding retained messages?

**Reply by zmstone** [reply]
HI

In EMQX Trie is for searching topics against wildcards, but not the other way around.

The retainer stores messages keyed by the topic name words(segments),
and there are index tables to speed up wildcard searches.

The indexed query is more efficient than searching the trie (if to implement using trie) from top down,
because there could be a very large number of matches at a higher level, but only one match going deeper.

**Reply by YYYYWD** [reply]
But if you receive a large number of retained messages, then the consistency consumption of maintaining such a trie tree should be very large, and you need to frequently update the trie tree in the kv storage, right?
